### PR TITLE
Editable: Adding the inlineToolbar prop

### LINKED ISSUE #541: Formatting text within Captions (inline format bar)
The formatting bar should appear only when you have text added in the caption field, not when you select the block.

### PR DESCRIPTION
**Pull request by youknowriad**
Some Editable need to show the formatting toolbar inlined with the Editable. This PR adds the `inlineToolbar` prop to solve this.

closes #541 

**Testing instructions**

 - Focus a text block, notice that the formatting toolbar is docked to the top of the block
 - Focus the caption of an image block and notice that the toolbar is shown Right on the top of the Editable.

### CODE REVIEW

**Comment by jasmussen** [comment]
I really love this. Looks super duper awesome! I'm impressed you all got this working!

There's one behavior that we need, but okay to do as a separate PR. Selecting the image block should select only that, and not set text focus anywhere. Like this:

image

The caption is surfaced as an editable area with palceholder text. When you then click the textarea, you get this:

captioning

**Comment by youknowriad** [comment]
Updated, focusing the image block don't focus the caption anymore. The caption is still shown even if it's empty and the block is not selected but this should be fixed by another PR.

**Comment by jasmussen** [comment]
👍

File: blocks/components/editable/index.js [hunk 1]
```diff
@@ -320,46 +326,62 @@ export default class Editable extends wp.element.Component {
...
 
 	render() {
-		const { tagName, style, value, focus, className, showAlignments = false, formattingControls } = this.props;
-		const classes = classnames( 'blocks-editable', className );
+		const {
+			tagName,
+			style,
+			value,
+			focus,
+			className,
```
Comment (aduth) [inline blocks/components/editable/index.js]: It's a little strange to me that when a parent component renders this one, the `className` is applied to a nested element, not the top-level element. Could we not just apply it to the `<div>` and update other styles to target the TinyMCE element?

File: blocks/components/editable/index.js [hunk 2]
```diff
@@ -112,11 +111,18 @@ export default class Editable extends wp.element.Component {
 	}
 
 	getRelativePosition( node ) {
-		const editorPosition = this.editor.getBody().closest( '.editor-visual-editor__block' ).getBoundingClientRect();
 		const position = node.getBoundingClientRect();
+
+		// Find the parent "relative" positioned container
+		const container = this.props.inlineToolbar
+			? this.editor.getBody().closest( '.blocks-editable' )
```
Comment (aduth) [inline blocks/components/editable/index.js]: Should we just assign a `ref` to the `.blocks-editable` node rendered by this component? I'd worried we were making more assumptions about surrounding content. I still think we should look to eliminate traversing to `.editor-visual-editor__block` but I'm not going to worry about it a whole lot here since it already exists.

File: blocks/library/image/index.js [hunk 3]
```diff
@@ -92,6 +92,8 @@ registerBlock( 'core/image', {
 			);
 		}
 
+		const focusCaption = ( focusValue = {} ) => setFocus( { editable: 'caption', ...focusValue } );
```
Comment (aduth) [inline blocks/library/image/index.js]: Object spread (or more accurately, `Object.assign`) will allow undefined values as long as it's not the first argument, so technically the `= {}` default isn't necessary.

File: blocks/components/editable/index.js (hunk 2 above)
Comment (youknowriad) [inline blocks/components/editable/index.js]: I agree. I'll play with this separately. I have some ideas to explore here.

**Comment by youknowriad** [comment]
Can I have an approval for this PR?

File: blocks/components/editable/index.js [hunk 4]
```diff
@@ -327,51 +334,59 @@ export default class Editable extends wp.element.Component {
...
+					<div className="block-editable__inline-toolbar">
+						{ formatToolbar }
+					</div>
+				}
+
+				<TinyMCE
+					tagName={ tagName }
+					onSetup={ this.onSetup }
+					style={ style }
+					className={ classes }
```
Comment (iseulde) [inline blocks/components/editable/index.js]: It's not clear with this component that changing the values later won't affect it. Should it accept classes and styles if that's not the case?

File: blocks/components/editable/format-toolbar.js [hunk 5]
```diff
@@ -150,7 +150,7 @@ class FormatToolbar extends wp.element.Component {
 				{ !! formats.link && ! this.state.isEditingLink &&
 					<div className="editable-format-toolbar__link-modal" style={ linkStyle }>
 						<a className="editable-format-toolbar__link-value" href="" onClick={ this.editLink }>
-							{ decodeURI( this.state.linkValue ) }
+							{ this.state.linkValue && decodeURI( this.state.linkValue ) }
```
Comment (iseulde) [inline blocks/components/editable/format-toolbar.js]: Is this related?

Comment (youknowriad) [inline blocks/components/editable/format-toolbar.js]: Not related, I was just seeing a "Null" showing when the link was empty

File: blocks/components/editable/index.js (hunk 4 above)
Comment (youknowriad) [inline blocks/components/editable/index.js]: Good catch! I just forgot to remove this when I moved the className handling to the `Editable` component.

File: blocks/components/editable/index.js [hunk 6]
```diff
@@ -112,11 +112,18 @@ export default class Editable extends wp.element.Component {
...
+			? this.editor.getBody().closest( '.blocks-editable' )
+			: this.editor.getBody().closest( '.editor-visual-editor__block' );
+		const editorPosition = container.getBoundingClientRect();
+
+		// The margins are different depending on the container
 		return {
-			top: position.top - editorPosition.top + 40 + ( position.height ),
-			left: position.left - editorPosition.left - 157
+			top: position.top - editorPosition.top + position.height + ( this.props.inlineToolbar ? 50 : 40 ),
+			left: position.left - editorPosition.left - ( this.props.inlineToolbar ? 100 : 157 )
```
Comment (iseulde) [inline blocks/components/editable/index.js]: These start looking like magic numbers. Could they be put in variables to make it clearer what they are? Can they not be calculated?

Comment (youknowriad) [inline blocks/components/editable/index.js]: I'm not sure how to calculate them. These numbers are related to the margins/paddings and left/right value of the container. I agree this is not ideal. I'll see if I can improve or fallback to constants for now.

Comment (youknowriad) [inline blocks/components/editable/index.js]: I've added some comments and constants. I hope it's better now.

Comment (iseulde) [inline blocks/components/editable/index.js]: I find it still quite hard to follow, and there's still a magic 40 and 50, but I maybe let's revisit that later then. :)

Comment (youknowriad) [inline blocks/components/editable/index.js]: Yes, the `40` and `50` are just the top values of the toolbar component (depending if it's inline or not). This is not perfect right! but it does the job right now. We can improve it later :)

**Review by ellatrix — APPROVED** [review]

File: blocks/components/editable/index.js [hunk 7]
```diff
@@ -118,12 +118,21 @@ export default class Editable extends wp.element.Component {
 		const container = this.props.inlineToolbar
 			? this.editor.getBody().closest( '.blocks-editable' )
 			: this.editor.getBody().closest( '.editor-visual-editor__block' );
-		const editorPosition = container.getBoundingClientRect();
+		const containerPosition = container.getBoundingClientRect();
+		const blockPadding = 14;
+		const blockMoverMargin = 18;
+
+		// These offsets are necessary because the toolbar where the link modal lives
+		// is absolute positionned and it's not shown when we compute the position here
```
Comment (paulwilde) [inline blocks/components/editable/index.js]: /s/positionned/positioned
